File: apps/go/domain/type_/fizz_buzz_type.go
// Package type_ contains FizzBuzz generation strategies.
package type_ //nolint:revive

import (
	"strconv"

	"github.com/k2works/getting-started-tdd/apps/go/domain/model"
)

// FizzBuzzType はタイプごとの FizzBuzz 生成を抽象化するインターフェースです。
type FizzBuzzType interface { //nolint:revive
	Generate(number int) model.FizzBuzzValue
}

// BasicGenerate は手続き型で FizzBuzz を生成します。
func BasicGenerate(number, fizzBuzzType int) string {
	isFizz := number%3 == 0
	isBuzz := number%5 == 0

	switch fizzBuzzType {
	case 1:
		if isFizz && isBuzz {
			return "FizzBuzz"
		}
		if isFizz {
			return "Fizz"
		}
		if isBuzz {
			return "Buzz"
		}
		return strconv.Itoa(number)
	case 2:
		return strconv.Itoa(number)
	case 3:
		if isFizz && isBuzz {
			return "FizzBuzz"
		}
		return strconv.Itoa(number)
	default:
		panic("該当するタイプは存在しません")
	}
}

// fizzBuzzTypeBase は FizzBuzz 判定の共通ロジックを提供します。
type fizzBuzzTypeBase struct{}

func (b fizzBuzzTypeBase) isFizz(number int) bool {
	return number%3 == 0
}

func (b fizzBuzzTypeBase) isBuzz(number int) bool {
	return number%5 == 0
}

// FizzBuzzType01 は通常の FizzBuzz を生成します。
type FizzBuzzType01 struct { //nolint:revive
	fizzBuzzTypeBase
}

// Generate はタイプ 1 の通常 FizzBuzz 値を返します。
func (f FizzBuzzType01) Generate(number int) model.FizzBuzzValue {
	if f.isFizz(number) && f.isBuzz(number) {
		return model.NewFizzBuzzValue(number, "FizzBuzz")
	}
	if f.isFizz(number) {
		return model.NewFizzBuzzValue(number, "Fizz")
	}
	if f.isBuzz(number) {
		return model.NewFizzBuzzValue(number, "Buzz")
	}
	return model.NewFizzBuzzValue(number, strconv.Itoa(number))
}

// FizzBuzzType02 は数値のみを返します。
type FizzBuzzType02 struct{} //nolint:revive

// Generate はタイプ 2 の数値文字列の FizzBuzz 値を返します。
func (f FizzBuzzType02) Generate(number int) model.FizzBuzzValue {
	return model.NewFizzBuzzValue(number, strconv.Itoa(number))
}

// FizzBuzzType03 は FizzBuzz のみ返し、それ以外は数値を返します。
type FizzBuzzType03 struct { //nolint:revive
	fizzBuzzTypeBase
}

// Generate はタイプ 3 の FizzBuzz または数値文字列の FizzBuzz 値を返します。
func (f FizzBuzzType03) Generate(number int) model.FizzBuzzValue {
	if f.isFizz(number) && f.isBuzz(number) {
		return model.NewFizzBuzzValue(number, "FizzBuzz")
	}
	return model.NewFizzBuzzValue(number, strconv.Itoa(number))
}

// NewFizzBuzzType は指定されたタイプの FizzBuzzType を生成します。
func NewFizzBuzzType(fizzBuzzType int) FizzBuzzType {
	switch fizzBuzzType {
	case 1:
		return FizzBuzzType01{}
	case 2:
		return FizzBuzzType02{}
	case 3:
		return FizzBuzzType03{}
	default:
		panic("該当するタイプは存在しません")
	}
}

File: apps/python/lib/fizzbuzz_type.py
from abc import ABC, abstractmethod


class FizzBuzzType(ABC):
    @staticmethod
    def create(type_: int) -> "FizzBuzzType":
        if type_ == 1:
            return FizzBuzzType01()
        if type_ == 2:
            return FizzBuzzType02()
        if type_ == 3:
            return FizzBuzzType03()
        raise ValueError("未定義のタイプです")

    @abstractmethod
    def generate(self, number: int) -> str:
        raise NotImplementedError

    def _is_fizz(self, number: int) -> bool:
        return number % 3 == 0

    def _is_buzz(self, number: int) -> bool:
        return number % 5 == 0

    def _is_fizz_buzz(self, number: int) -> bool:
        return number % 15 == 0


class FizzBuzzType01(FizzBuzzType):
    def generate(self, number: int) -> str:
        if self._is_fizz_buzz(number):
            return "FizzBuzz"
        if self._is_fizz(number):
            return "Fizz"
        if self._is_buzz(number):
            return "Buzz"
        return str(number)


class FizzBuzzType02(FizzBuzzType):
    def generate(self, number: int) -> str:
        return str(number)


class FizzBuzzType03(FizzBuzzType):
    def generate(self, number: int) -> str:
        if self._is_fizz_buzz(number):
            return "FizzBuzz"
        return str(number)
